src/coordinator/tsdb/remote/compressed_codecs.go

```diff
@@ -0,0 +1,236 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package remote
+
+import (
+	"io"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/generated/proto/rpc"
+
+	"github.com/m3db/m3db/src/dbnode/encoding"
+	"github.com/m3db/m3db/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3db/src/dbnode/ts"
+	"github.com/m3db/m3db/src/dbnode/x/xio"
+	"github.com/m3db/m3db/src/dbnode/x/xpool"
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+	xtime "github.com/m3db/m3x/time"
+)
+
+// RPCFromSeriesIterator converts a seriesIterator to a grpc series
+func RPCFromSeriesIterator(it encoding.SeriesIterator) (*rpc.Series, error) {
+	defer it.Close()
+
+	replicas := it.Replicas()
+	compressedReplicas := make([]*rpc.CompressedValuesReplica, 0, len(replicas))
+	for _, replica := range replicas {
+		replicaSegments := make([]*rpc.Segments, 0, len(replicas))
+		readers := replica.Readers()
+		next := true
+		for next {
+			segments := &rpc.Segments{}
+			l, _, _ := readers.CurrentReaders()
+			if l == 1 {
+				br := readers.CurrentReaderAt(0)
+				segment, err := br.Segment()
+				if err != nil {
+					return nil, err
+				}
+				segments.Merged = &rpc.Segment{
+					Head: segment.Head.Bytes(),
+					Tail: segment.Tail.Bytes(),
+				}
+			} else {
+				unmerged := make([]*rpc.Segment, 0, l)
+				for i := 0; i < l; i++ {
+					br := readers.CurrentReaderAt(i)
+					segment, err := br.Segment()
+					if err != nil {
+						return nil, err
+					}
+					unmerged = append(unmerged, &rpc.Segment{
+						Head: segment.Head.Bytes(),
+						Tail: segment.Tail.Bytes(),
+					})
+				}
+				segments.Unmerged = unmerged
+			}
+			replicaSegments = append(replicaSegments, segments)
+			next = readers.Next()
+		}
+		compressedReplicas = append(compressedReplicas, &rpc.CompressedValuesReplica{
+			Segments: replicaSegments,
+		})
+	}
+
+	start := xtime.ToNanoseconds(it.Start())
+	end := xtime.ToNanoseconds(it.End())
+
+	compressedDatapoints := &rpc.CompressedDatapoints{
+		Namespace: it.Namespace().String(),
+		StartTime: start,
+		EndTime:   end,
+		Replicas:  compressedReplicas,
+	}
+
+	tagIter := it.Tags()
+	tags := make(map[string]string)
+	for tagIter.Next() {
```

Comment:
maybe throw this into a separate function? could be useful elsewhere?

Reply:
I think if we need it in the future, we can pull it out then. This one converts to `map[string]string` which isn't necessarily a type we'd want to use elsewhere though, as we tend to want to convert to more useful types.

That being said, might change it from `map[string]string` to a list of structures for slight perf boost